xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,785 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+			b.intervalTimer.Stop()
+		}
+		b.intervalTimer = afterFunc(interval, func() {
+			b.intervalTimerAlgorithm()
+		})
+	} else {
+		// "If a config is provided with both the `success_rate_ejection` and
+		// `failure_percentage_ejection` fields unset, skip starting the timer and
+		// unset the timer start timestamp."
+		b.timerStartTime = time.Time{}
+		// Should we stop the timer here as well? Not defined in gRFC but I feel
+		// like it might make sense as you don't want to eject addresses. Also
+		// how will addresses eventually get unejected in this case if only one
+		// more pass of the interval timer after no-op configuration comes in?
+	}
+	b.mu.Unlock()
+	b.pickerUpdateCh.Put(lbCfg)
+
+	// then pass the address list along to the child policy.
+	b.childMu.Lock()
+	defer b.childMu.Unlock()
+	return b.child.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState:  s.ResolverState,
+		BalancerConfig: b.cfg.ChildPolicy.Config,
+	})
+}
+
+func (b *outlierDetectionBalancer) ResolverError(err error) {
+	if b.child != nil {
+		b.childMu.Lock()
+		defer b.childMu.Unlock()
+		b.child.ResolverError(err)
+	}
+}
+
+func (b *outlierDetectionBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	scw, ok := b.scWrappers[sc]
+	if !ok {
+		// Return, shouldn't happen if passed up scw
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(b.scWrappers, scw.SubConn)
+	}
+	b.scUpdateCh.Put(&scUpdate{
+		scw:   scw,
+		state: state,
+	})
+
+}
+
+func (b *outlierDetectionBalancer) Close() {
+	b.closed.Fire()
+	if b.child != nil {
+		b.childMu.Lock()
+		b.child.Close()
+		b.child = nil
+		b.childMu.Unlock()
+	}
+
+	// Any other cleanup needs to happen (subconns, other resources?)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.intervalTimer != nil {
+		b.intervalTimer.Stop()
+	}
+}
+
+func (b *outlierDetectionBalancer) ExitIdle() {
+	if b.child == nil {
+		return
+	}
+	if ei, ok := b.child.(balancer.ExitIdler); ok {
+		b.childMu.Lock()
+		defer b.childMu.Unlock()
+		ei.ExitIdle()
+		return
+	}
+
+	// Fallback for children handled in clusterimpl balancer Removing SubConns
+	// is defined in API and also in graceful switch balancer, but already done
+	// in ClusterImpl. I guess we should do that here?
+}
+
+// wrappedPicker delegates to the child policy's picker, and when the request
+// finishes, it increments the corresponding counter in the map entry referenced
+// by the subConnWrapper that was picked.
+type wrappedPicker struct {
+	childPicker balancer.Picker
+	noopPicker  bool
+}
+
+func (wp *wrappedPicker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	pr, err := wp.childPicker.Pick(info)
+	if err != nil {
+		return balancer.PickResult{}, err
+	}
+
+	done := func(di balancer.DoneInfo) {
+		if !wp.noopPicker {
+			incrementCounter(pr.SubConn, di)
+		}
+		if pr.Done != nil {
+			pr.Done(di)
+		}
+	}
+	// Shouldn't happen, defensive programming.
```

Comment:
Please log an error for these so they'll show up and fail tests.

Reply:
Wait actually this can happen, adding this error failed my test. If the child sends a picker which picks a nil SubConn. (i.e. a TestConstPicker without a set SubConn, this check will fail). This is a valid scenario right? Thus, I deleted my comment and the logger.Errorf() call.

Reply:
Whoops, missed this one. Will do, thanks.